src/local_scheduler/local_scheduler.cc

```diff
@@ -663,22 +664,21 @@ void reconstruct_task_update_callback(Task *task,
 
   /* Otherwise, the test-and-set succeeded, so resubmit the task for execution
    * to ensure that reconstruction will happen. */
-  TaskSpec *spec = Task_task_spec(task);
+  TaskExecutionSpec *execution_spec = Task_task_execution_spec(task);
+  TaskSpec *spec = execution_spec->Spec();
   if (ActorID_equal(TaskSpec_actor_id(spec), NIL_ACTOR_ID)) {
-    handle_task_submitted(state, state->algorithm_state, Task_task_spec(task),
-                          Task_task_spec_size(task));
+    handle_task_submitted(state, state->algorithm_state, *execution_spec);
   } else {
-    handle_actor_task_submitted(state, state->algorithm_state,
-                                Task_task_spec(task),
-                                Task_task_spec_size(task));
+    handle_actor_task_submitted(state, state->algorithm_state, *execution_spec);
   }
 
   /* Recursively reconstruct the task's inputs, if necessary. */
-  for (int64_t i = 0; i < TaskSpec_num_args(spec); ++i) {
-    int count = TaskSpec_arg_id_count(spec, i);
+  int64_t num_dependencies = execution_spec->NumDependencies();
+  for (int64_t i = 0; i < num_dependencies; ++i) {
```

Comment:
I'm a little confused about this code. Why do we need nested for loops? Don't we just want to loop over the object ID dependencies in the args and then loop over the execution dependencies?

Reply:
It's to match the way the task args are structured, where each argument index may actually have 0 to multiple object IDs. The dependencies of `execution_spec` include both the task arguments and the execution dependencies.